data_structures/linked_list/circular_linked_list.py

```diff
@@ -2,201 +2,137 @@
-        >>> print(f"{len(cll)}: {cll}")
-        2: <Node data=1> => <Node data=2>
-        >>> cll.delete_rear()
-        >>> print(f"{len(cll)}: {cll}")
-        1: <Node data=1>
-        >>> cll.delete_rear()
-        >>> print(f"{len(cll)}: {cll}")
-        0: Empty linked list
-        """
-        if not self.head:
-            raise IndexError("Deleting from an empty list")
-
-        temp_node, current_node = self.head, self.head
-
-        if current_node.next_ptr == current_node:
-            self.head = None
+            temp = self.head
+            for _ in range(index - 1):
+                temp = temp.next
+            new_node.next = temp.next
+            temp.next = new_node
+            if index == len(self) - 1:  # insert at tail
+                self.tail = new_node
+
+    def delete_front(self):
+        return self.delete_nth(0)
+
+    def delete_tail(self) -> None:
+        return self.delete_nth(len(self) - 1)
+
+    def delete_nth(self, index: int = 0):
+        if not 0 <= index < len(self):
+            raise IndexError("list index out of range.")
+        delete_node = self.head
+        if self.head == self.tail:  # just one node
+            self.head = self.tail = None
+        elif index == 0:  # delete head node
+            self.tail.next = self.tail.next.next
+            self.head = self.head.next
         else:
-            while current_node.next_ptr != self.head:
-                temp_node = current_node
-                current_node = current_node.next_ptr
+            temp = self.head
+            for _ in range(index - 1):
+                temp = temp.next
+            delete_node = temp.next
+            temp.next = temp.next.next
+            if index == len(self) - 1:  # delete at tail
+                self.tail = temp
+        return delete_node.data
 
-            temp_node.next_ptr = current_node.next_ptr
+    def print_list(self) -> None:
+        print(self)
 
-        self.length -= 1
-        if not self.head:
-            assert self.length == 0
+    def is_empty(self):
+        return len(self) == 0
+
+
+def test_circular_linked_list() -> None:
+    """
+    >>> test_circular_linked_list()
+    """
+    circular_linked_list = CircularLinkedList()
+    assert len(circular_linked_list) == 0
+    assert circular_linked_list.is_empty() is True
+    assert str(circular_linked_list) == ""
+
+    try:
+        circular_linked_list.delete_front()
+        assert False  # This should not happen
+    except IndexError:
+        assert True  # This should happen
+
+    try:
+        circular_linked_list.delete_tail()
+        assert False  # This should not happen
+    except IndexError:
+        assert True  # This should happen
+
+    try:
+        circular_linked_list.delete_nth(-1)
+        assert False
+    except IndexError:
+        assert True
+
+    try:
+        circular_linked_list.delete_nth(0)
+        assert False
+    except IndexError:
+        assert True
+
+    for i in range(5):
```

Comment:
```suggestion
    assert circular_linked_list.is_empty() is True
    for i in range(5):
```

Reply:
No.  We are looking for a test where .is_empty() returns False.  We already have one that returns True.